Fixing json output html

TornaxO7 opened the pull request:
Should fix [this issue](https://github.com/TornaxO7/himalaya/issues/5).

TornaxO7 committed:
Fixing JSON output

- JSON of message includes `hasAttachment` key now
- JSON output shows both body types: Text and Html
- Changed `Body` struct so it can store html and text now.

TornaxO7 committed:
Tests

Updated tests with the latest Body implementation

TornaxO7 committed:
Merge branch 'mail_refactor' of github.com:TornaxO7/himalaya into fixing_json_output_html

TornaxO7 commented:
Summary: 
- JSON Output includes `hasAttachment` now
- JSON output gives `text/html` and `text/plain` back

TornaxO7 commented:
@soywod review pls

soywod commented on src/msg/body.rs:
```diff
@@ -24,91 +23,130 @@ error_chain! {
-        &mut self.0
-    }
-}
+        let content = if let Some(text) = self.text.clone() {
```
This means that the `text` version of the body will always be taken first. In fact, it should show the text or the html depending on this argument https://github.com/soywod/himalaya/blob/608057b23bcd067e6f5fe130c6e395655432a448/src/msg/cli.rs#L112-L119.

TornaxO7 replied:
Oh yes, you're right. Thank you for pointing this out!

TornaxO7 replied:
Hm... ok so in my latest commit, I'm printing out everything of the message: Body + Headers. Do you think that this is fine?

soywod replied:
It depends for what, I think both should be available: a simple version with only the body, and a full version with headers + body.

TornaxO7 replied:
So I did the following in my latest commmit:
- The `Display` trait of `Msg` is gonna return the _body_ of the message.
- I added another method: `get_full_message` which adds the headers as well.

Do you acgree with this?

soywod replied:
Sounds great :ok_hand:

soywod commented on src/msg/model.rs:
```diff
@@ -36,9 +35,8 @@ use colorful::Colorful;
-        // An error appeared, when it tried to parse the body of the msg!
         ParseBody (err: String) {
-            description("Couldn't get the body of the parsed msg."),
+            description("An error appeared, when it tried to parse the body of the msg!"),
```
Try to align with other errors (I mean, the style of the message).

soywod commented on src/msg/model.rs:
```diff
@@ -275,14 +271,14 @@ impl Msg {
-        let mut new_body = self
-            .body
+        let mut new_body = self.body.text.clone()
+            .unwrap_or(String::new())
```
`unwrap_or_default` should work here?

TornaxO7 replied:
yes thanks for this hint!

soywod commented on src/msg/model.rs:
```diff
@@ -728,6 +732,23 @@ impl Msg {
+    where
+        S: ser::Serializer
+    {
+        println!("ÖLKJASFÖLKJ");
```
:thinking:

TornaxO7 replied:
*cough cough* eeeeeh.... I'll fix this *cough cough* (it was used to be a quick log message)

soywod replied:
:rofl:

soywod commented on src/msg/model.rs:
```diff
@@ -728,6 +732,23 @@ impl Msg {
 }
 
 // -- Traits --
+impl ser::Serialize for Msg {
```
Isn't possible to have a dedicated struct to avoid implementing the Serialize trait? See https://github.com/soywod/himalaya/issues/153.

TornaxO7 replied:
The reason why I implemented it manually, is because I didn't want to add another attribute for `Msg` which is only used for Serializsation. I'll add this as a comment/doc to the trait implementation.

soywod replied:
I still think it could be nicer to introduce a dedicated struct for this purpose, sth like:

```rust
#[derive(Serialize)]
struct SerializableMsg {
    #[serde(flatten)]
    inner: Msg,
    has_attachments: Bool,
}
```

It's better to let the `serde` lib do it's job rather than implementing ourselves Serialize.

TornaxO7 replied:
Btu that makes it more complicated from my point of view. Currently, we can just give `Msg` to the `Output.print` function and the method will decide if it should use the implemented serializer or the `Display` trait.

How should it be handled if you want to print out the msg? I'd need to always check which mode the printer has `Output.is_json()` and `Output.is_text()`.

soywod replied:
I'm not sure to understand your problem. Instead of passing a `Msg` to the printer, you pass a `SerializableMsg`. The only thing we need is a connector from a `Msg` to a `SerializableMsg`. It's quite common to have different struct from the same model for different purposes. At one of my work I used [`diesel`](https://diesel.rs/guides/getting-started) (it's an ORM), and it's common to have a struct for the `Queryable` trait, another one for the `Insertable` trait etc.

TornaxO7 replied:
Could you please show my a little example what you mean? Should he `SerializeableMsg` implement the `From` trait to be used like this?
```
Output.print(SerializeableMsg::from(&msg));
```
?

soywod replied:
It could be great but I'm afraid you will not be able (because you will miss the `has_attachments` that does not exists on `Msg`). A custom function will be needed.

soywod reviewed (changes requested):
I think there is still an issue with the formatting, could you check if the config file is taken into consideration by your IDE?

TornaxO7 committed:
Fixes

- Removed suspicious println macro in serializer of msg... *cough cough*
- Fixed output in the "read" command
- othe small fixes

TornaxO7 committed:
Formatting

Formatted all files

TornaxO7 commented:
@soywod update

TornaxO7 committed:
Msg

- Adding 'get_full_message' method which prints out all information of the
  message in a string

TornaxO7 committed:
New Msg-Struct

Adding MsgSerialized, a struct, which represents the "correct" serialized
version of a message because it includes another attribute: `has_attachment`.

TornaxO7 committed:
Cleanup

Removed the manual serialize implementation of `Msg` and added a little more
info about the MsgSerialized.

TornaxO7 commented:
@soywod review pls

soywod commented on src/msg/model.rs:
```diff
@@ -846,18 +859,18 @@ impl TryFrom<&Fetch> for Msg {
-                    None
-                },
-            };
+                // the empty array represents an invalid body, so we can enter the
```
Tab issue?

soywod commented on src/msg/model.rs:
```diff
@@ -70,9 +69,39 @@ error_chain::error_chain! {
+
+impl From<&Msg> for MsgSerialized {
+    fn from(msg: &Msg) -> Self {
+        let has_attachment = msg.attachments.len() > 0;
```
It's a detail, but I would go with a more dedicated method `is_empty()`.

TornaxO7 replied:
Good idea!

TornaxO7 committed:
Little changes

- Used a better condition for checking if the message includes attachments or
  not
- format fixes

TornaxO7 commented:
@soywod update
